buildSrc/src/main/kotlin/zoneInfosResourcesGenerator.kt

```diff
@@ -0,0 +1,102 @@
+import java.io.File
+
+/*
+ * Copyright 2019-2024 JetBrains s.r.o. and contributors.
+ * Use of this source code is governed by the Apache 2.0 License that can be found in the LICENSE.txt file.
+ */
+
+private val pkg = "package kotlinx.datetime.timezones.tzData"
+
+private fun generateByteArrayProperty(tzData: TzData, header: String, propertyName: String): String = buildString {
+    append(header)
+    appendLine()
+    appendLine()
+    appendLine("/* ${tzData.fullTzNames.joinToString(", ")} */")
+    append("internal val $propertyName get() = byteArrayOf(")
+    for (chunk in tzData.data.toList().chunked(16)) {
+        appendLine()
+        append("    ")
+        val chunkText = chunk.joinToString {
+            it.toString().padStart(4, ' ')
+        } + ","
+        append(chunkText)
+    }
+    appendLine()
+    append(")")
+}
+
+private class TzData(val data: ByteArray, val fullTzNames: MutableList<String>)
+private fun loadTzBinaries(
+    zoneInfo: File,
+    currentName: String,
+    result: MutableList<TzData>
+) {
+    val zoneName = if (currentName.isEmpty()) zoneInfo.name else "$currentName/${zoneInfo.name}"
+    if (zoneInfo.isDirectory) {
+        zoneInfo.listFiles()?.forEach {
+            loadTzBinaries(it, zoneName, result)
+        }
+    } else {
+        val bytes = zoneInfo.readBytes()
+        val foundTzData = result.firstOrNull { it.data.contentEquals(bytes) }
+        val tzData: TzData
+        if (foundTzData != null) {
+            tzData = foundTzData
+        } else {
+            tzData = TzData(bytes, mutableListOf())
+            result.add(tzData)
+        }
+
+        tzData.fullTzNames.add(zoneName)
+    }
+}
+
+fun generateZoneInfosResources(zoneInfoDir: File, outputDir: File, version: String) {
+    if (!zoneInfoDir.exists() || !zoneInfoDir.isDirectory) error("zoneinfo does not exists")
+
+    val header = buildString {
+        appendLine()
+        append("/* AUTOGENERATED FROM ZONE INFO DATABASE v.$version */")
+        appendLine()
+        appendLine()
+        append(pkg)
+    }
+
+    val loadedZones = mutableListOf<TzData>()
+    zoneInfoDir.listFiles()?.forEach { file ->
+        loadTzBinaries(file, "", loadedZones)
+    }
+
+    val zoneDataByNameBody = StringBuilder()
+    val getTimeZonesBody = StringBuilder()
+    loadedZones.forEachIndexed { id, tzData ->
+        val tzDataName = "tzData$id"
+        val data = generateByteArrayProperty(tzData, header, tzDataName)
+        File(outputDir, "$tzDataName.kt").writeText(data)
+        tzData.fullTzNames.forEach { name ->
+            zoneDataByNameBody.appendLine("    \"$name\" -> $tzDataName")
+            getTimeZonesBody.appendLine("    \"$name\",")
+        }
+    }
+
+    val content = buildString {
+        append(header)
+        appendLine()
+        appendLine()
+        appendLine("internal fun zoneDataByName(name: String): ByteArray = when(name) {")
```

Comment:
Seems extremely suboptimal to compare 600 strings just to get to `CET`. Luckily, we have the trie implementation that was recently introduced to kotlinx-datetime. I've implemented some code to do this, here:
```kotlin
private class TrieNode<T>(
    val children: MutableList<Pair<String, TrieNode<T>>> = mutableListOf(),
    var terminalValue: T? = null
)

internal fun String.toKotlinCode(): String = buildString {
    append('"')
    for (c in this@toKotlinCode) {
        when (c) {
            '"' -> append("\\\"")
            '\\' -> append("\\\\")
            '\b' -> append("\\b")
            '\t' -> append("\\t")
            '\n' -> append("\\n")
            '\r' -> append("\\r")
            else -> append(c)
        }
    }
    append('"')
}

internal fun Char.toKotlinCode(): String = when (this) {
    '\'' -> "'\\''"
    else -> "'$this'"
}

fun<V> Map<String, V>.toTrieAsKotlinCode(valueToKotlinCode: (V) -> String): String {
    val trie = TrieNode<V>()
    for ((string, v) in this) {
        var node = trie
        for (char in string) {
            val searchResult = node.children.binarySearchBy(char.toString()) { it.first }
            node = if (searchResult < 0) {
                TrieNode<V>().also { node.children.add(-searchResult - 1, char.toString() to it) }
            } else {
                node.children[searchResult].second
            }
        }
        check(node.terminalValue == null)
        node.terminalValue = v
    }
    fun reduceTrie(trie: TrieNode<V>) {
        for ((_, child) in trie.children) {
            reduceTrie(child)
        }
        val newChildren = mutableListOf<Pair<String, TrieNode<V>>>()
        for ((key, child) in trie.children) {
            if (child.terminalValue == null && child.children.size == 1) {
                val (grandChildKey, grandChild) = child.children.single()
                newChildren.add(key + grandChildKey to grandChild)
            } else {
                newChildren.add(key to child)
            }
        }
        trie.children.clear()
        trie.children.addAll(newChildren.sortedBy { it.first })
    }
    reduceTrie(trie)
    val indent = "    "
    fun StringBuilder.printTrie(trie: TrieNode<V>, inputVar: String, index: Int = 0, currentIndent: String = "    ") {
        val terminalValue = trie.terminalValue
        if (terminalValue != null) {
            append("${currentIndent}if ($inputVar.length == $index) ${valueToKotlinCode(terminalValue)} else ")
        } else {
            append(currentIndent)
        }
        if (trie.children.isEmpty()) {
            appendLine("null")
        } else {
            appendLine("when {")
            for ((key, child) in trie.children) {
                val check = if (key.length == 1) {
                    "$inputVar[$index] == ${key[0].toKotlinCode()}"
                } else {
                    "$inputVar.startsWith(${key.toKotlinCode()}, startIndex = $index)"
                }
                appendLine("${currentIndent + indent}$check -> {")
                printTrie(child, inputVar, index + key.length, currentIndent + indent + indent)
                appendLine("${currentIndent + indent}}")
            }
            appendLine("${currentIndent + indent}else -> null")
            appendLine("${currentIndent}}")
        }
    }
    return buildString { printTrie(trie, "input") }
}

// test
fun main() {
    val strings = ZoneId.getAvailableZoneIds()
    strings.associateWith { it }.toTrieAsKotlinCode { "ZoneId.of(${it.toKotlinCode()})" }.also { println(it) }
}
```

Here's a sample of what this outputs:
```kotlin
    when {
        input[0] == 'A' -> {
            when {
                input.startsWith("frica/", startIndex = 1) -> {
                    when {
                        input[7] == 'A' -> {
                            when {
                                input.startsWith("bidjan", startIndex = 8) -> {
                                    if (input.length == 14) ZoneId.of("Africa/Abidjan") else null
                                }
                                input.startsWith("ccra", startIndex = 8) -> {
                                    if (input.length == 12) ZoneId.of("Africa/Accra") else null
                                }

```

Reply:
Probably makes more sense to output the code immediately, though, and not store it as a string like I did.

Reply:
Oh, is that so? Cool, then there isn't much point.